connector/linkedin/linkedin.go

```diff
@@ -149,15 +157,46 @@ func (p profile) fullname() string {
 	return fname
 }
 
+func (c *linkedInConnector) primaryEmail(ctx context.Context, client *http.Client) (email string, err error) {
+	req, err := http.NewRequest("GET", apiURL+"/emailAddress?q=members&projection=(elements*(handle~))", nil)
+	if err != nil {
+		return email, fmt.Errorf("new req: %v", err)
+	}
+
+	resp, err := client.Do(req.WithContext(ctx))
+	if err != nil {
+		return email, fmt.Errorf("get URL %v", err)
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return email, fmt.Errorf("read body: %v", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return email, fmt.Errorf("%s: %s", resp.Status, body)
+	}
+
+	var parsedResp emailresp
+	err = json.Unmarshal(body, &parsedResp)
+	if err == nil {
+		for _, elem := range parsedResp.Elements {
+			email = elem.Handle.EmailAddress
+		}
+	}
```

Comment:
❓ Is this validation good enough? From glancing at the API docs (thanks for including the links, much appreciated!), [`handle~` could be missing](https://docs.microsoft.com/en-us/linkedin/shared/integrations/people/primary-contact-api#sample-decoration-error) because something (no idea what!) went wrong...?

With the current code, that would mean `email` remains `""` (unassigned) and `err` would still be `nil` because of the way JSON unmarshaling works. (Its errors aren't for cases where expected fields aren't found, but rather more low-level stuff. If you try to unmarshal `{this:"garba"ge}}` for example.)

💡 looks like we'd catch that below, [in this check](https://github.com/dexidp/dex/pull/1460/files#diff-306cce14d2c51d1b284c11a81280c3acR225) -- I think I'd propose moving that into this helper method: "if `primaryEmail` returns a non-nil error, it has successfully retrieved the email address" sounds like a decent property for a function with that name. What do you think?

Reply:
Yes, agree. Will make that change in a bit.